test/helpers/URMHandlerBase.sol

```diff
@@ -0,0 +1,467 @@
+///      - _getDescription() - to return the description string
+abstract contract URMHandlerBase is Test {
+  using LibRollbackSet for RollbackSet;
+
+  FakeProtocolContract[] public targets;
+  RollbackSet public _rollbackSet;
+  bytes4[] public selectors;
+
+  address public admin;
+  address public guardian;
+
+  // Call tracking
+  mapping(bytes32 => uint256) public calls;
+  uint256 public ghost_rollbackExistsReverts;
+  uint256 public ghost_invalidOperationReverts;
+  uint256 public ghost_authorizationReverts;
+  uint256 public ghost_unableToFindProposals;
+
+  modifier countCall(bytes32 key) {
+    calls[key]++;
+    _;
+  }
+
+  constructor(address _admin, address _guardian, FakeProtocolContract[] memory _targets) {
+    admin = _admin;
+    guardian = _guardian;
+
+    for (uint256 i = 0; i < _targets.length; i++) {
+      targets.push(_targets[i]);
+    }
+
+    selectors = [FakeProtocolContract.setFee.selector, FakeProtocolContract.setFeeGuardian.selector];
+
+    // Note: URM address will be set by child class after initialization
+  }
+
+  /// @notice Abstract method to get the URM contract instance
+  /// @return The URM contract instance
+  function _getURM() internal view virtual returns (URMCore);
+
+  /// @notice Abstract method to get the description string
+  /// @return The description string
+  function _getDescription() internal view virtual returns (string memory);
+
+  /// @notice Set the URM address in the RollbackSet (called by child constructor)
+  function _setURMAddress() internal {
+    _rollbackSet.setURM(address(_getURM()));
+  }
+
+  /// @notice Propose a rollback
+  /// @param _rollbackFee The new fee
+  function propose(uint256 _rollbackFee) external countCall("propose") {
+    // Get the rollback transactions
+    (address[] memory _targets, uint256[] memory _values, bytes[] memory _calldatas) =
+      RollbackTransactionGenerator.generateRandomRollbackTransactions(_rollbackFee, guardian, targets, selectors);
+
+    // Get the rollback ID
+    uint256 _rollbackId = _getURM().getRollbackId(_targets, _values, _calldatas, _getDescription());
+
+    // Only propose if it doesn't already exist
+    if (_rollbackSet.contains(_rollbackId)) {
+      ghost_rollbackExistsReverts++;
+      return; // Skip instead of reverting
+    }
+
+    vm.prank(admin);
+    _getURM().propose(_targets, _values, _calldatas, _getDescription());
+
+    _rollbackSet.add(
+      RollbackProposal({
+        targets: _targets,
+        values: _values,
+        calldatas: _calldatas,
+        description: _getDescription(),
+        rollbackId: _rollbackId
+      })
+    );
+  }
+
+  /// @notice Queue a valid rollback proposal
+  /// @param _randomIndex Used to randomly select a queueable proposal
+  function queue(uint256 _randomIndex) external countCall("queue") {
+    // Only proceed if there are pending proposals
+    if (!_rollbackSet.hasProposalsInState(IGovernor.ProposalState.Pending)) {
```

Comment:
nit: just an idea, could be DRYed into a `_onlyContinueIfStateFoundInSet(IGovernor.ProposalState.Pending)`

Reply:
i was thinking it looks like this:

```
function _onlyContinueIfStateFoundInSet(ProposalState memory _proposalState) internal {
   if (!_rollbackSet.hasProposalsInState(_proposalState)) {
      ghost_unableToFindProposals++;
      return;
   }
}
```

usage:

```
_onlyContinueIfStateFoundInSet(IGovernor.ProposalState.Pending);
```

Reply:
```javascript
function _onlyContinueIfStateFoundInSet(ProposalState memory _proposalState) internal {
   if (!_rollbackSet.hasProposalsInState(_proposalState)) {
      ghost_unableToFindProposals++;
      return; // <-- returns from THIS helper function only
   }
}

function queue(uint256 _randomIndex) external countCall("queue") {
    _onlyContinueIfStateFoundInSet(IGovernor.ProposalState.Pending);

    //  this WILL still run even if above check fails and it returns 
    _doSomething();
}
```

Reply:
oh,  you are very correct